ide/projectapi/src/org/netbeans/modules/projectapi/SimpleFileOwnerQueryImplementation.java

```diff
@@ -414,6 +421,18 @@ private static URI goUp(URI u) {
         assert u.toString().startsWith(nue.toString()) : "not a parent: " + nue + " of " + u;
         return nue;
     }
+
+    private static Set<String> extractFolderPaths(String folders, String pathSeparator) {
+        Set<String> paths = Arrays.stream(folders.split(pathSeparator))   //NOI18N
+                .filter(s -> !s.trim().isEmpty())
+                .map(f -> FileUtil.toFileObject(FileUtil.normalizeFile(new File(f))))
+                .filter(Objects::nonNull)
+                .map(FileObject::getPath)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+        return paths.isEmpty() ? Collections.emptySet() : paths.size() == 1 ? Collections.singleton(paths.iterator().next()) : paths;
```

Comment:
> Not quite sure about this cascade - is there a reason why this should improve the behavior?
> 
I can change the code to the imperative style using `for` and `if`. The cascade was not meant for improving behaviour.

> I guess I'd say something like: `return paths.isEmpty() ? null : paths;` should work well enough? (Using the short-circuit behavior when the value is `null`, if we didn't want to use that, I would say simply `return paths`.)
> 
- I was unsure if returning `null` from the utility method would be frowned upon here. :)
- The reason for returning `Collections.singleton(paths)` is to be more performant during the comparison; for the usual scenario i.e. of a single root path configured).

Please let me know if I should just change it to `return paths;` 
Thanks.

Reply:
I know that writing streams is modern and trendy ... but please think twice before coding such a cascade. If there's any issue, the code like this is literally undebuggable.

Streams processing should be IMHO only used for larger number of items where internal logic could eventually parallelize/speed up the processing; not instread of old-school but readable cycles and conditions.

Reply:
Understood 😄 Thank you @lahodaj and @sdedic.

Reply:
just as sidenote: I don't actually mind streams like that, what I do however is to filter early and to combine filters.

so instead of:
```
map(..).filter(nonNull).map(..).filter(nonNull)...
```
you would get:
```
.filter(foo != null && foo.getOther() != null)
.map(::getOther)
```
this is often easier to read (for java devs) since it is very similar to how the body of a for-loop would look like.

regarding debugging: NB is lacking good support for stream debugging, but `peek()` or intermediate temporary cuts with `toList()` to debug through the stream works just fine in a lot of cases. Sure you have to change the code slightly before setting a breakpoint but its often not a big deal. (exceptions do never combine very well with streams though thats when the loops always win)